### ISSUE TITLE
Rack does not build against correct openssl library on Windows

### ISSUE THREAD

**Issue by cschol** [VCVRack/Rack#822]
Operating system(s): Windows 10
Source revision: 617236be0666a1051d50bcc096e2ca5f4fcb5554

While trying to troubleshoot with a post-mortem Visual Studio debugger I built the `dist` version of Rack.

I tried to run Rack from the `dist/Rack` directory by doubleclicking `Rack.exe` in the Explorer and get this error:

![ssl_error_rack_0 6](https://user-images.githubusercontent.com/624427/37636838-e6ef640e-2bc9-11e8-8b8a-b0b74aa3eb1b.png)

**Update**: launching Rack from the `msys2` console in the same directory works fine.

**Comment by cschol** [comment]
I think this is a problem with `libcurl` and `openssl`.

```
$ nm libcurl-4.dll | grep SSL_COMP
000000007084e24c I __imp_SSL_COMP_free_compression_methods
0000000070839740 T SSL_COMP_free_compression_methods
```

```
$ nm libssl-1_1-x64.dll | grep SSL_COMP
000000006d496b50 T SSL_COMP_add_compression_method
000000006d496af0 T SSL_COMP_get_compression_methods
000000006d496d30 T SSL_COMP_get_id
000000006d496d00 T SSL_COMP_get_name
000000006d496d20 T SSL_COMP_get0_name
000000006d496b20 T SSL_COMP_set0_compression_methods
```

`SSL_COMP_free_compression_methods` was deprecated in this [openssl commit](https://github.com/openssl/openssl/commit/03b0e735556ed). Rack is using a very recent v1.1.0g.

I also have `openssl` 1.0.2 installed on this system as part of `msys2`. I wonder if this is why the build process works, but when Rack is run with `openssl` 1.1.0 from the `dist` outside of `msys2` it does not.

Trying to set `OPENSSL_API_COMPAT` to 1.0.2 to see if that fixes it, see [here](https://wiki.openssl.org/index.php/OpenSSL_1.1.0_Changes).

**Comment by cschol** [comment]
Aha. From `./config` output:

```
configure: Added /mingw64/lib to LD_LIBRARY_PATH
checking for OpenSSL headers version... 1.0.2 - 0x100020efL
checking for OpenSSL library version... 1.0.2
checking for OpenSSL headers and library versions matching... yes
```

It is **not** using 1.1.0 that is provided by Rack in the `dep` directory, but the system (msys2) version of `openssl`. Investigating.

**Comment by cschol** [comment]
Eureka!

```
configure: Added /home/myuser/src/Rack-cschol/dep/lib to LD_LIBRARY_PATH
checking for OpenSSL headers version... 1.1.0 - 0x1010007fL
checking for OpenSSL library version... 1.1.0
checking for OpenSSL headers and library versions matching... yes
```

Needed to update line 120 of `dep/Makefile` to set the specific `PKG_CONFIG_PATH`:

```
cd curl-7.56.0 && PKG_CONFIG_PATH="$(LOCAL)/lib/pkgconfig" ./configure --prefix="$(LOCAL)" \
```

See this [post](https://askubuntu.com/questions/475670/how-to-build-curl-with-the-latest-openssl) for reference.

Now Rack starts up without error when running it from the `dist` directory outside of `msys2`.

@AndrewBelt I can't test this change on a Mac, but on Windows and Linux it works correctly (`./configure ` output shows it is using the correct path on both platforms).

**Comment by cschol** [comment]
@AndrewBelt I was trying to figure out if this was only a problem on my system. Do you see the correct libraries for openssl getting linked when doing make dep on Windows to build libcurl?

**Comment by AndrewBelt** [comment]
Haven't tested thoroughly, but it seems to be working fine on my Windows build machine. Will test the official build on another machine and solve this later if it's not a problem with mine.

**Comment by cschol** [comment]
OK. Thanks. I recently did an `msys2` upgrade (via `pacman`). Might be related to that.

**Comment by dizzisound** [comment]
@cschol @AndrewBelt On my Windows 8.1 system, after building a master dist, I got exactly the same _Entry Point_ error. In my case too, I could conversely launch Rack from the `msys2/mingw64` shell. 
I tried some diagnosis, and I found that the `libcurl-4.dll` library dinamically loads a couple of openssl dependencies: `libeay32.dll` and `ssleay32`. 
When launching from the msys2 environment, `libcurl `points to the correct version (1.0.2.14) of the dll's, located in my `mingw64/bin` path, and that's fine. 
Conversely, when running the Rack executable -- both the version regularly installed in my Program Files folder (using the installer generated with `makensis`), or the Rack executable located in the dist folder of my build -- directly from the Windows GUI or the command prompt, the 2 openssl dependencies are retrieved scanning through the Windows system PATH variable. Now, it happens that existing system versions (1.0.0.7) of the 2 dll's are detected (the ones going with an Intel iCLS Client installed on my machine) and these instances are referenced, resulting in a loading error.
For immediate use, I can easily solve this removing the Intel client path from my system path, though this is quite rude. Otherwise, and better, if I copy the correct dll's (i.e. the `mingw64 `versions) in the Rack folder, I can safely run Rack from a normal Explorer window or shortcut. Obviously, these workarounds are not the case for a distributed release.
Hope this can give some hints for sorting out the issue at the build script level, should it become a verified, repeatable issue reported from testing.
If needed, I can replicate the issue and upload some screenshots or logs.
In the meantime, forgive me for my English full of issues!
Best

**Comment by dizzisound** [comment]
**Update**: Put aside the the anomaly in my environment PATH setup, some more ispectioning revealed that my issues are due to my build. In fact I verified that my build generated a faulty `libcurl.dll` library, which references the mismatched dll's. Still I couldn't try a new build with the latest sources and makefiles, but I can assume the issues have been overcome, by this time.
BTW, I downloaded and tested the Rack 0.6 pre-release as per Andrew FB post here: https://www.facebook.com/groups/vcvrack/permalink/168305840496235/ 
and it works fine; both on my DAW, equipped with a Windows 7 64-bit OS, and on my dev machine, equipped with a Windows 8.1 version, Rack launches with no issues.
Ultimately my note can be held for expired.
Best
